### PR TITLE
Do not generate "inc/dec [mem]" when asked to generate debuggable code.

### PR DESCRIPTION
**Pull request by sivarv**
With PR https://github.com/dotnet/coreclr/pull/8101, RyuJIT started generating inc/dec [mem] for RMW write operations that add/sub 1 from a memory operand.  But amd64 debugger routine NativeWalker::Decode() has hit an assert while decoding modR/M byte of "inc dword ptr[rax]" instruction.  It it not clear whether this routine can correctly handle all possible mem addr modes with inc/dec instruction.  For this reason, inc/dec [mem] is not generated while generating debuggable code.

### CODE REVIEW

File: src/jit/lowerxarch.cpp [hunk 1]
```diff
@@ -3968,8 +3968,22 @@ bool Lowering::IsRMWMemOpRootedAtStoreInd(GenTreePtr tree, GenTreePtr* outIndirC
     *outIndirCandidate = nullptr;
     *outIndirOpSource  = nullptr;
 
-    // Early out if storeInd is already known to be a non-RMW memory op
     GenTreeStoreInd* storeInd = tree->AsStoreInd();
+
+    // Do not perform RMW pattern check in the following modes:
+    //   1) MinOgetMaxIntrinsicSIMDVectorLengthpts - so that it serves as a bail out route in case of
```
Comment (jkotas) [inline src/jit/lowerxarch.cpp]: MinOgetMaxIntrinsicSIMDVectorLengthpts  ?

Comment (sivarv) [inline src/jit/lowerxarch.cpp]: That was an unexpected typo.  Corrected it.

**Comment by sivarv** [comment]
@dotnet/jit-contrib - Please review this.

**Comment by CarolEidt** [comment]
Do I understand correctly that:
1. This is intended to address the decoding issue with inc/dec
2. This disables the RMW transformation for *all* ops?

If so, it seems to be a big hammer for the inc/dec issue.

**Comment by sivarv** [comment]
@CarolEidt - Yes your understanding is right that I intend to address decoding issue with inc/dec.  To keep the things simple even in debugging scenario, I took this opportunity to disable RMW optimization for debuggable code.

**Comment by CarolEidt** [comment]
@sivarv - I'm not sure that's the right tradeoff. I believe that the legacy jits did this kind of address "optimization" (what I would consider just code generation, rather than optimization) in MinOpts. While it is reasonable to be conservative with MinOpts, I think we should focus on eliminating:
- transformations that are costly, and
- transformations that are somewhat "risky" (admittedly hard to nail down)
I wouldn't say that the RMW transformations meet either of those criteria. The fact that there was a decoding issue doesn't seem to justify this.

**Comment by sivarv** [comment]
@CarolEidt - you are right that  Legacy jit32 (x86) still generates inc/dec [mem] even in minopts/dbgCode.
But legacy Jit64 didn't generate inc/dec [mem] when asked to generate dbgCode.

I have updated my fix not to generate inc/dec [mem] when asked to generate debuggable code till issue #8139 is fixed.  With this fix desktop debugger test on amd64 passes.

**Comment by CarolEidt** [comment]
Awesome, thanks! I think that's the right solution.

**Comment by sivarv** [comment]
@dotnet-bot retest Ubuntu x64 Checked Build and Test

**Comment by sivarv** [comment]
@dotnet-bot Ubuntu x64 Checked Build and Test

**Comment by sivarv** [comment]
@dotnet-bot retest Ubuntu x64 Checked

**Comment by RussKeldorph** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test

**Pull request merged**